File: dener8/programmers_가장큰수.py
'''
처음 푼 방식 => 오답
'''
# def solution(numbers):
#     answer = ''

#     max_num = max(numbers)
#     max_length = len(str(max_num))
#     for i in range(len(numbers)):
#         numbers[i] = [int(str(numbers[i]) + str(numbers[i])[-1] * (max_length - len(str(numbers[i])))), len(str(numbers[i]))]
#     numbers = sorted(numbers, key = lambda x: [-x[0], x[1]])

#     for number in numbers:
#         answer += str(number[0])[ : number[1]]

#     answer = str(int(answer))

#     return answer

'''
두 번째 => 정답
'''

def solution(numbers):
    answer = ''
    numbers = sorted(numbers, key=lambda x: str(x) * 3, reverse=True)  # padding을 해서 정렬 (실제 값은 바뀌지 않음)

    if numbers[0] == 0:
        return '0'

    answer = ''.join(map(str, numbers))

    return answer

File: myunghyunnero/programmers_가장큰수.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;
bool cmp(string a, string b){
    return a+b > b+a;
}
string solution(vector<int> numbers) {
    string answer = "";
    vector<string> strs;
    for(int number : numbers){
        strs.push_back(to_string(number));
    }
    sort(strs.begin(),strs.end(),cmp);
    for(string str : strs){
        answer+=str;
    }
    if(strs[0]=="0"){
        return "0";
    }
    return answer;
}